store/subscription.go

```diff
@@ -0,0 +1,316 @@
+		}
+
+		// Construct the requested addresses
+		contractsAddrs := make(map[gethCommon.Address]map[gethCommon.Address]struct{}, len(s.erc20List))
+
+		// Add ETH contract addresses
+		contractsAddrs[model.ETHAddress] = make(map[gethCommon.Address]struct{}, len(newSubs))
+		for addr := range newSubs {
+			contractsAddrs[model.ETHAddress][addr] = struct{}{}
+		}
+
+		// Add ERC20 contract addresses
+		for addr := range s.erc20List {
+			erc20Addr := gethCommon.HexToAddress(addr)
+			contractsAddrs[erc20Addr] = make(map[gethCommon.Address]struct{}, len(newSubs))
+			for addr := range newSubs {
+				contractsAddrs[erc20Addr][addr] = struct{}{}
+			}
+		}
+
+		// Construct the erc20 contract addresses
+		s.newBalances, err = s.balancer.BalanceOf(ctx, blockNumber, contractsAddrs)
+		if err != nil {
+			s.logger.Error("Failed to get ERC20 balance", "len", len(s.newBalances), "err", err)
+			return err
+		}
+
+		// Update ERC20 balances
+		for token, addrs := range s.newBalances {
+			for addr, balance := range addrs {
+				err = s.accountStore.InsertAccount(&model.Account{
+					ContractAddress: token.Bytes(),
+					BlockNumber:     s.blockNumber,
+					Address:         addr.Bytes(),
+					Balance:         balance.String(),
+				})
+				if err != nil {
+					s.logger.Error("Failed to insert ERC20 account", "err", err)
+					return err
+				}
+			}
+		}
+
+		// Update subscription table
+		for _, sub := range subs {
+			sub.BlockNumber = s.blockNumber
+			err = s.subscriptionStore.UpdateBlockNumber(sub)
+			if err != nil {
+				s.logger.Error("Failed to update block number", "err", err)
+				return err
+			}
+		}
+		// Construct the new subs
+		s.newSubs = make(map[gethCommon.Address]*model.Subscription)
+		for _, sub := range subs {
+			s.newSubs[gethCommon.BytesToAddress(sub.Address)] = sub
+		}
+
+		return nil
+	}
+
+	return nil
+}
+
+func (s *subscription) insert(ctx context.Context, events []*model.Transfer) error {
+	mapAddrs := make(map[string]struct{})
+	// Collect modified addresses
+	for _, e := range events {
+		mapAddrs[common.BytesToHex(e.From)] = struct{}{}
+		mapAddrs[common.BytesToHex(e.To)] = struct{}{}
+	}
+	var addrs [][]byte
+	for addr := range mapAddrs {
+		addrs = append(addrs, common.HexToBytes(addr))
+	}
+
+	// Check whether it's a subscribed account
+	subs, err := s.subscriptionStore.FindAddresses(addrs)
+	if err != nil {
+		s.logger.Error("Failed to find subscription address", "len", len(addrs), "err", err)
+		return err
+	}
+
+	// Empty subscriptions, ignore it
+	if len(subs) == 0 {
+		s.logger.Info("No subscribed accounts found")
+		return nil
+	}
+
+	addrDiff := make(map[gethCommon.Address]map[gethCommon.Address]*big.Int)
+	contractsAddrs := make(map[gethCommon.Address]map[gethCommon.Address]struct{})
+	// Insert events if it's a subscribed accounts
+	for _, e := range events {
+		for _, sub := range subs {
+			if bytes.Equal(e.From, sub.Address) || bytes.Equal(e.To, sub.Address) {
```

Comment:
if !bytes.Equal(e.From, sub.Address) && !bytes.Equal(e.To, sub.Address) {
return
}

Reply:
map optimized